File: back/DevC/src/main/java/com/back/devc/global/security/jwt/AuthCookieService.java
package com.back.devc.global.security.jwt;

import jakarta.servlet.http.HttpServletResponse;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseCookie;
import org.springframework.stereotype.Component;

@Component
public class AuthCookieService {

    @Value("${custom.jwt.access-cookie-name:access_token}")
    private String accessCookieName;

    @Value("${custom.jwt.access-cookie-secure:false}")
    private boolean accessCookieSecure;

    @Value("${custom.jwt.access-cookie-same-site:Lax}")
    private String accessCookieSameSite;

    @Value("${custom.jwt.access-token-expiration-seconds:3600}")
    private long accessTokenExpirationSeconds;

    // 로그인 성공 시 Access Token을 HttpOnly Cookie로 응답에 추가
    // 쿠키 만료 시간은 설정값(custom.jwt.access-token-expiration-seconds)을 사용
    public void addAccessTokenCookie(HttpServletResponse response, String token) {
        setAccessTokenCookie(response, token, accessTokenExpirationSeconds);
    }

    // 로그아웃 시 Access Token 쿠키를 즉시 만료
    // 빈 값과 maxAge 0을 내려 브라우저가 기존 쿠키를 제거
    public void expireAccessTokenCookie(HttpServletResponse response) {
        setAccessTokenCookie(response, "", 0);
    }

    // Access Token 쿠키를 생성해 Set-Cookie 헤더로 응답에 추가
    // HttpOnly, Secure, SameSite, Path, Max-Age 등 쿠키 정책을 한 곳에서 관리
    public void setAccessTokenCookie(HttpServletResponse response, String token, long maxAgeSeconds) {
        ResponseCookie cookie = ResponseCookie.from(accessCookieName, token == null ? "" : token)
                .httpOnly(true)
                .secure(accessCookieSecure)
                .path("/")
                .maxAge(maxAgeSeconds)
                .sameSite(accessCookieSameSite)
                .build();

        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString());
    }

    // Access Token 만료 쿠키 헤더 문자열을 생성
    public String buildExpiredAccessTokenCookieHeader() {
        return ResponseCookie.from(accessCookieName, "")
                .httpOnly(true)
                .secure(accessCookieSecure)
                .path("/")
                .maxAge(0)
                .sameSite(accessCookieSameSite)
                .build()
                .toString();
    }
}

File: back/DevC/src/main/kotlin/com/back/devc/global/security/jwt/AuthCookieService.kt
package com.back.devc.global.security.jwt

import jakarta.servlet.http.HttpServletResponse
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpHeaders
import org.springframework.http.ResponseCookie
import org.springframework.stereotype.Component

@Component
class AuthCookieService(
    @Value("\${custom.jwt.access-cookie-name:access_token}")
    private val accessCookieName: String,

    @Value("\${custom.jwt.access-cookie-secure:false}")
    private val accessCookieSecure: Boolean,

    @Value("\${custom.jwt.access-cookie-same-site:Lax}")
    private val accessCookieSameSite: String,

    @Value("\${custom.jwt.access-token-expiration-seconds:3600}")
    private val accessTokenExpirationSeconds: Long
) {

    // 로그인 성공 시 Access Token을 HttpOnly Cookie로 응답에 추가한다.
    fun setAccessTokenCookie(response: HttpServletResponse, token: String?) {
        addAccessTokenCookie(response, token, accessTokenExpirationSeconds)
    }

    // 로그아웃 또는 회원탈퇴 시 Access Token 쿠키를 즉시 만료시킨다.
    fun expireAccessTokenCookie(response: HttpServletResponse) {
        addAccessTokenCookie(response, "", 0)
    }

    // Access Token 쿠키를 Set-Cookie 헤더로 응답에 추가
    // 실제 쿠키 생성은 buildAccessTokenCookie()에 위임
    private fun addAccessTokenCookie(
        response: HttpServletResponse,
        token: String?,
        maxAgeSeconds: Long
    ) {
        response.addHeader(HttpHeaders.SET_COOKIE, buildAccessTokenCookie(token, maxAgeSeconds))
    }

    // Access Token 쿠키 문자열을 생성
    // HttpOnly, Secure, Path, Max-Age, SameSite 등 쿠키 정책을 한 곳에서 관리
    private fun buildAccessTokenCookie(token: String?, maxAgeSeconds: Long): String {
        return ResponseCookie.from(accessCookieName, token ?: "")
            .httpOnly(true)
            .secure(accessCookieSecure)
            .path("/")
            .maxAge(maxAgeSeconds)
            .sameSite(accessCookieSameSite)
            .build()
            .toString()
    }
}